[Bug] hs_persist_state silently ignores unknown variable names

## Summary

Passing a name that does not correspond to any variable produces an empty state string with no warning or error.

## Current behavior
-  returns success with no output.
- Callers cannot distinguish a typo from an intentional omission.
 
## Expected behavior
- Detect unknown variable names and fail with a clear error (or emit a warning).

## Severity
Medium.

## Notes
- This is currently covered by tests added for unknown names.

## Revised Assessment (Task 2)

Agreed — a typo in a variable name list is a bug; a warning would hide it. Changing to a hard stopping error.

**Revised fix:**
- Add `readonly HS_ERR_UNKNOWN_VAR_NAME=10` to the public error codes.
- In the `hs_persist_state_as_code` loop: if `declare -p "$var" >/dev/null 2>&1` fails (variable not declared in scope), emit `[ERROR] hs_persist_state_as_code: '$name' is not declared in scope.` and `return "$HS_ERR_UNKNOWN_VAR_NAME"` (fail-fast before writing anything).
- Declared-but-unset locals continue to be skipped silently (no change).

**Updated files:**
- `config/handle_state.sh`: new error code `HS_ERR_UNKNOWN_VAR_NAME=10`; updated loop.
- `test/test-hs_persist_state.bats`:
  - `known issue nr.1` (line 728): remove `known_issue` tag; `run -$HS_ERR_UNKNOWN_VAR_NAME --separate-stderr`; assert `$stderr` contains the error message.
  - `hs_persist_state_as_code does not include variables that were not set in init` (line 714): split into two tests — one for a set variable (`foo`) and one for `bar` which is undeclared in the call stack (now an error).
- `docs/libraries/handle_state.rst`: remove "unknown variable names" Known Limitation; add `HS_ERR_UNKNOWN_VAR_NAME=10` to error codes list and to the `hs_persist_state_as_code` Errors section.

**No blockers.** Awaiting your approval.

## Implementation Plan (Task 3)

**Branch:** `bug/issue-1-error-on-undeclared-var`

### config/handle_state.sh

1. Add `readonly HS_ERR_UNKNOWN_VAR_NAME=10` after `HS_ERR_INVALID_ARGUMENT_TYPE=9`.
2. In `hs_persist_state_as_code`, in the per-variable loop (currently lines 147–176), wrap the existing `if [ "${!__var_name+x}" ]` with an outer `declare -p` check:
   ```bash
   if ! declare -p "$__var_name" >/dev/null 2>&1; then
       echo "[ERROR] hs_persist_state_as_code: '$__var_name' is not declared in scope." >&2
       return "$HS_ERR_UNKNOWN_VAR_NAME"
   fi
   # existing set/unset logic follows
   ```
3. Update the function's `# Errors:` comment block to document `HS_ERR_UNKNOWN_VAR_NAME`.

### docs/libraries/handle_state.rst

1. Add `HS_ERR_UNKNOWN_VAR_NAME=10` to the Error Codes list.
2. Add `HS_ERR_UNKNOWN_VAR_NAME=10` to the `hs_persist_state_as_code` Errors section.
3. Remove "unknown variable names passed to `hs_persist_state_as_code` are ignored" from Known Limitations.
4. Update the Behavior bullet "Unknown names and function names are ignored" to reflect the new error.

### test/test-hs_persist_state.bats (preliminary tests only — Task 4)

- Update `known issue nr.1` (line 728): remove `known_issue` tag, assert `run -$HS_ERR_UNKNOWN_VAR_NAME` and error message on stderr.
- Split line 714 test: the existing test keeps `foo` (set), adds a separate test asserting that passing undeclared `bar` from `init` now errors.

### Testing strategy
Run `bats test/test-hs_persist_state.bats` locally. New tests must fail before implementation (xfail pattern), pass after.

--- a/.vscode/settings.json
+++ b/.vscode/settings.json
@@ -15,5 +15,6 @@
         "release/*"
     ],
     "git.branchProtectionPrompt": "always",
-    "files.readonlyFromPermissions": true
+    "files.readonlyFromPermissions": true,
+    "shellcheck.useWorkspaceRootAsCwd": false
 }

--- a/config/handle_state.sh
+++ b/config/handle_state.sh
@@ -154,8 +154,12 @@ hs_persist_state_as_code() {
             return "$HS_ERR_RESERVED_VAR_NAME"
         fi
         # Fail fast if the name is not declared anywhere in the dynamic scope
-        # (catches typos and function names passed by mistake).
+        # (catches typos). Function names are silently skipped — persisting
+        # functions is tracked separately as a known limitation.
         if ! declare -p "$__var_name" >/dev/null 2>&1; then
+            if declare -f "$__var_name" >/dev/null 2>&1; then
+                continue
+            fi
             echo "[ERROR] hs_persist_state_as_code: '$__var_name' is not declared in scope." >&2
             return "$HS_ERR_UNKNOWN_VAR_NAME"
         fi

--- a/config/handle_state.sh
+++ b/config/handle_state.sh
@@ -45,6 +45,7 @@ readonly HS_ERR_VAR_NAME_NOT_IN_STATE=6
 readonly HS_ERR_STATE_VAR_UNINITIALIZED=7
 readonly HS_ERR_MISSING_ARGUMENT=8
 readonly HS_ERR_INVALID_ARGUMENT_TYPE=9
+readonly HS_ERR_UNKNOWN_VAR_NAME=10
 
 # --- hs_persist_state_as_code ----------------------------------------------------------
 # Function:
@@ -78,6 +79,8 @@ readonly HS_ERR_INVALID_ARGUMENT_TYPE=9
 #     already defined in the prior state object.
 #   - `HS_ERR_CORRUPT_STATE` if the prior state object cannot be evaluated
 #     safely during collision checking.
+#   - `HS_ERR_UNKNOWN_VAR_NAME` if a requested variable name is not declared
+#     in the caller's scope (catches typos and function names).
 # Usage examples:
 #   local state_var
 #   init() {
@@ -150,11 +153,14 @@ hs_persist_state_as_code() {
             echo "[ERROR] hs_persist_state_as_code: refusing to persist reserved variable name '$__var_name'." >&2  
             return "$HS_ERR_RESERVED_VAR_NAME"
         fi
-        # Check if the variable exists in the caller (local or global). We avoid
-        # using `local -p` here because that only inspects locals of this
-        # function, not the caller's scope. If the variable exists, capture its
-        # value and emit a guarded assignment that will only set it in the
-        # receiving scope if that scope has declared it `local`.
+        # Fail fast if the name is not declared anywhere in the dynamic scope
+        # (catches typos and function names passed by mistake).
+        if ! declare -p "$__var_name" >/dev/null 2>&1; then
+            echo "[ERROR] hs_persist_state_as_code: '$__var_name' is not declared in scope." >&2
+            return "$HS_ERR_UNKNOWN_VAR_NAME"
+        fi
+        # The variable is declared. If it is set, persist it; if unset, skip
+        # silently — an unset local is a legitimate intentional omission.
         if [ "${!__var_name+x}" ]; then
             # Get the value of the variable
             local var_value

--- a/config/handle_state.sh
+++ b/config/handle_state.sh
@@ -415,8 +415,7 @@ hs_read_persisted_state() {
             _hs_read_requested_state_vars() {
                 local __state=$1
                 local __quiet_mode=$2
-                shift
-                shift
+                shift 2
                 local __requested_var
                 local __restored_entries=""
 

--- a/docs/libraries/handle_state.rst
+++ b/docs/libraries/handle_state.rst
@@ -68,7 +68,8 @@ variables to the opaque state object named by ``-S``.
 Behavior:
 
 - Requested variables that are unset are skipped.
-- Unknown names and function names are ignored.
+- Variable names not declared in the caller's scope are an error.
+- Function names passed as variable names are an error.
 - Indexed arrays currently persist only their first element.
 - Associative arrays are ignored.
 - Namerefs are persisted as scalar values.
@@ -86,6 +87,8 @@ Errors:
   the prior state.
 - ``HS_ERR_CORRUPT_STATE=4``: the prior state could not be evaluated safely
   during collision checking.
+- ``HS_ERR_UNKNOWN_VAR_NAME=10``: a requested variable name is not declared
+  in the caller's scope.
 
 hs_destroy_state
 ~~~~~~~~~~~~~~~~
@@ -243,14 +246,13 @@ Error Codes
 - ``HS_ERR_STATE_VAR_UNINITIALIZED=7``
 - ``HS_ERR_MISSING_ARGUMENT=8``
 - ``HS_ERR_INVALID_ARGUMENT_TYPE=9``
+- ``HS_ERR_UNKNOWN_VAR_NAME=10``
 
 Known Limitations
 -----------------
 
 The tests currently demonstrate these limitations:
 
-- unknown variable names passed to ``hs_persist_state_as_code`` are ignored
-- function names passed to ``hs_persist_state_as_code`` are ignored
 - indexed arrays preserve only their first element
 - associative arrays are ignored
 - namerefs are restored as scalar values

--- a/docs/libraries/handle_state.rst
+++ b/docs/libraries/handle_state.rst
@@ -69,7 +69,7 @@ Behavior:
 
 - Requested variables that are unset are skipped.
 - Variable names not declared in the caller's scope are an error.
-- Function names passed as variable names are an error.
+- Function names are ignored.
 - Indexed arrays currently persist only their first element.
 - Associative arrays are ignored.
 - Namerefs are persisted as scalar values.
@@ -253,6 +253,7 @@ Known Limitations
 
 The tests currently demonstrate these limitations:
 
+- function names passed to ``hs_persist_state_as_code`` are ignored
 - indexed arrays preserve only their first element
 - associative arrays are ignored
 - namerefs are restored as scalar values

--- a/test/test-hs_persist_state.bats
+++ b/test/test-hs_persist_state.bats
@@ -715,8 +715,8 @@ fi'
   # shellcheck disable=SC2329
   f() {
     local state_var=""
-    init(){ local foo=one bar; hs_persist_state_as_code -S "$1" foo bar; }
-    cleanup(){ local foo bar; hs_read_persisted_state -S "$1" -- foo bar; printf "%s:%s" "$foo" "${bar:-}"; }
+    init(){ local foo=one bar; hs_persist_state_as_code "$@" -- foo bar; }
+    cleanup(){ local foo bar; hs_read_persisted_state -q "$@" -- foo bar; printf "%s:%s" "$foo" "${bar:-}"; }
     init -S state_var
     cleanup -S state_var
   }
@@ -730,7 +730,7 @@ fi'
   # shellcheck disable=SC2329
   f() {
     local state_var=""
-    init(){ hs_persist_state_as_code -S "$1" not_a_var; }
+    init(){ hs_persist_state_as_code "$@" -- not_a_var; }
     init -S state_var
   }
   run -"$HS_ERR_UNKNOWN_VAR_NAME" --separate-stderr f
@@ -742,7 +742,7 @@ fi'
   # shellcheck disable=SC2329
   f() {
     local state_var=""
-    init(){ my_func(){ echo "nope"; }; hs_persist_state_as_code -S "$1" my_func; }
+    init(){ my_func(){ echo "nope"; }; hs_persist_state_as_code "$@" -- my_func; }
     init -S state_var
   }
   run -"$HS_ERR_UNKNOWN_VAR_NAME" --separate-stderr f

--- a/test/test-hs_persist_state.bats
+++ b/test/test-hs_persist_state.bats
@@ -715,8 +715,8 @@ fi'
   # shellcheck disable=SC2329
   f() {
     local state_var=""
-    init(){ local foo=one bar; hs_persist_state_as_code "$@" -- foo bar; }
-    cleanup(){ local foo bar; hs_read_persisted_state -q "$@" -- foo bar; printf "%s:%s" "$foo" "${bar:-}"; }
+    init(){ local foo=one unset_var; hs_persist_state_as_code "$@" -- foo unset_var; }
+    cleanup(){ local foo unset_var; hs_read_persisted_state -q "$@" -- foo unset_var; printf "%s:%s" "$foo" "${unset_var:-}"; }
     init -S state_var
     cleanup -S state_var
   }
@@ -826,11 +826,11 @@ fi'
 @test "preserve special characters in persisted values" {
   # shellcheck disable=SC2329
   f() {
-    init(){ local foo='a b "c" $d'; hs_persist_state_as_code -S "$1" foo; }
-    cleanup(){ local foo; eval "$1"; printf "%s" "$foo"; }
-    state=""
-    init state
-    cleanup "$state"
+    init(){ local foo="a b \"c\" \$d"; hs_persist_state_as_code "$@" -- foo; }
+    cleanup(){ local foo; hs_read_persisted_state "$@" -- foo; printf "%s" "$foo"; }
+    local state=""
+    init -S state
+    cleanup -S state
   }
   run -0 f
   [ "$output" = "a b \"c\" \$d" ]
@@ -853,17 +853,14 @@ fi'
       local foo=two
       hs_persist_state_as_code bad -S "$1" -- foo
     }
-    cleanup() {
-      local foo=""
-      eval "$state"
-      printf "%s" "$foo"
-    }
-    state=""
+    local state=""
     init state
-    cleanup
+    local -a names=()
+    _hs_extract_persisted_state_var_names f "$state" names
+    [[ " ${names[*]} " != *" bad "* ]]
+    [[ " ${names[*]} " == *" foo "* ]]
   }
   run -0 --separate-stderr f
-  [ "$output" = "two" ]
   [ -z "$stderr" ]
 }
 
@@ -989,16 +986,17 @@ fi'
 @test "hs_persist_state_as_code with -S var_name assigns to variable" {
   # shellcheck disable=SC2329
   f() {
+    local encoded=""
     init() {
       local bar=two
       hs_persist_state_as_code -S "$1" bar
     }
     cleanup(){
       local bar
-      eval "$state"
+      eval "$encoded"
       printf "%s" "$bar"
     }
-    init state
+    init encoded
     cleanup
   }
   run -0 f
@@ -1025,18 +1023,18 @@ fi'
 @test "hs_destroy_state with -S rewrites the named variable in place" {
   # shellcheck disable=SC2329
   f() {
+    local rewritten=""
     init() {
       local foo=one bar=two
       hs_persist_state_as_code -S "$1" foo bar
     }
     cleanup() {
       local foo="" bar=""
-      eval "$state"
+      eval "$rewritten"
       printf "%s:%s" "${foo:-}" "${bar:-}"
     }
-    state=""
-    init state
-    hs_destroy_state -S state foo
+    init rewritten
+    hs_destroy_state -S rewritten foo
     cleanup
   }
   run -0 --separate-stderr f
@@ -1046,6 +1044,7 @@ fi'
 
 # bats test_tags=hs_destroy_state
 @test "hs_destroy_state rebuild works in a clean shell without exported helper functions" {
+  # shellcheck disable=SC2016
   run -0 --separate-stderr env -i PATH="$PATH" LIB="$LIB" bash --noprofile -lc '
     source "$LIB"
     init() {
@@ -1093,18 +1092,18 @@ fi"
 @test "hs_destroy_state ignores forwarded args before final --" {
   # shellcheck disable=SC2329
   f() {
+    local trimmed=""
     init() {
       local foo=one bar=two
       hs_persist_state_as_code -S "$1" foo bar
     }
     cleanup() {
       local foo="" bar=""
-      eval "$state"
+      eval "$trimmed"
       printf "%s:%s" "${foo:-}" "${bar:-}"
     }
-    state=""
-    init state
-    hs_destroy_state bad -S state -- foo
+    init trimmed
+    hs_destroy_state bad -S trimmed -- foo
     cleanup
   }
   run -0 --separate-stderr f

--- a/test/test-hs_persist_state.bats
+++ b/test/test-hs_persist_state.bats
@@ -737,16 +737,18 @@ fi'
   [[ "$stderr" == *"'not_a_var' is not declared in scope"* ]]
 }
 
-# bats test_tags=hs_persist_state_as_code
-@test "hs_persist_state_as_code errors when a function name is passed instead of a variable" {
+# bats test_tags=hs_persist_state_as_code,known_issue
+@test "known issue nr.2: hs_persist_state_as_code silently ignores function names" {
   # shellcheck disable=SC2329
   f() {
+    my_func(){ echo "nope"; }
     local state_var=""
-    init(){ my_func(){ echo "nope"; }; hs_persist_state_as_code "$@" -- my_func; }
-    init -S state_var
+    hs_persist_state_as_code -S state_var my_func
+    printf "%s" "$state_var"
   }
-  run -"$HS_ERR_UNKNOWN_VAR_NAME" --separate-stderr f
-  [[ "$stderr" == *"'my_func' is not declared in scope"* ]]
+  run -0 --separate-stderr f
+  [ -z "$output" ]
+  [ -z "$stderr" ]
 }
 
 # bats test_tags=hs_persist_state_as_code,known_issue
@@ -1144,7 +1146,7 @@ fi"
 @test "known issue nr.6: hs_destroy_state does not reject internal state variable name __state_var_names explicitly" {
   # shellcheck disable=SC2329
   f() {
-    local __state_var_names='not a state snippet'
+    local -a __state_var_names=('not a state snippet')
     hs_destroy_state -S __state_var_names foo >/dev/null
   }
   run -"$HS_ERR_CORRUPT_STATE" --separate-stderr f

--- a/test/test-hs_persist_state.bats
+++ b/test/test-hs_persist_state.bats
@@ -711,45 +711,42 @@ fi'
 }
 
 # bats test_tags=hs_persist_state_as_code
-@test "hs_persist_state_as_code does not include variables that were not set in init" {
+@test "hs_persist_state_as_code persists a set variable and skips a declared-but-unset one" {
   # shellcheck disable=SC2329
   f() {
-    init(){ local foo=one; hs_persist_state_as_code -S "$1" foo bar; }
-    cleanup(){ local foo bar; eval "$1"; printf "%s:%s" "$foo" "${bar:-}"; }
-    state=""
-    init state
-    cleanup "$state"
+    local state_var=""
+    init(){ local foo=one bar; hs_persist_state_as_code -S "$1" foo bar; }
+    cleanup(){ local foo bar; hs_read_persisted_state -S "$1" -- foo bar; printf "%s:%s" "$foo" "${bar:-}"; }
+    init -S state_var
+    cleanup -S state_var
   }
-  run -0 f
+  run -0 --separate-stderr f
   [ "$output" = "one:" ]
+  [ -z "$stderr" ]
 }
 
-# bats test_tags=hs_persist_state_as_code,known_issue
-@test "known issue nr.1: hs_persist_state_as_code silently ignores unknown variable names" {
+# bats test_tags=hs_persist_state_as_code
+@test "hs_persist_state_as_code errors on a variable name not declared in scope" {
   # shellcheck disable=SC2329
   f() {
-    state=""
+    local state_var=""
     init(){ hs_persist_state_as_code -S "$1" not_a_var; }
-    init state
-    printf "%s" "$state"
+    init -S state_var
   }
-  run -0 --separate-stderr f
-  [ -z "$output" ]
-  [ -z "$stderr" ]
+  run -"$HS_ERR_UNKNOWN_VAR_NAME" --separate-stderr f
+  [[ "$stderr" == *"'not_a_var' is not declared in scope"* ]]
 }
 
-# bats test_tags=hs_persist_state_as_code,known_issue
-@test "known issue nr.2: hs_persist_state_as_code silently ignores function names" {
+# bats test_tags=hs_persist_state_as_code
+@test "hs_persist_state_as_code errors when a function name is passed instead of a variable" {
   # shellcheck disable=SC2329
   f() {
-    state=""
+    local state_var=""
     init(){ my_func(){ echo "nope"; }; hs_persist_state_as_code -S "$1" my_func; }
-    init state
-    printf "%s" "$state"
+    init -S state_var
   }
-  run -0 --separate-stderr f
-  [ -z "$output" ]
-  [ -z "$stderr" ]
+  run -"$HS_ERR_UNKNOWN_VAR_NAME" --separate-stderr f
+  [[ "$stderr" == *"'my_func' is not declared in scope"* ]]
 }
 
 # bats test_tags=hs_persist_state_as_code,known_issue

